### ISSUE TITLE
Subscription rules match on events which it shouldn't match on

### ISSUE THREAD

**Issue by jgoyvaerts** [Particular/NServiceBus.AzureServiceBus#703]
Right now we have 2 events, ISentMail and ISentMailWithDocument . 

- ISentMail has 3 different endpoints which subscribe to this event.
- ISentMailWithDocument has 1 endpoint which subscribes to to this event

As it is now, whenever we publish an ISentMailWithDocument event, it is sent to all 3 endpoints, because it matches the rule 
`[NServiceBus.EnclosedMessageTypes] LIKE 'Events.ISentMail%'`

I realise this particular (😄) example can be fixed by changing the namespace of the event, or making sure they don't have similar starting string, but I was just wondering in general if you consider this a bug or design decision which we have to deal with.

**Comment by SeanFeldman** [comment]
Historically, implementation was using `'Namespace.EventName'` as filter with 4 options:
1. event **is exactly** 'Namespace.EventName'`
1. event **starts with** 'Namespace.EventName'`
1. event **ends with** 'Namespace.EventName'`
1. event **contains** 'Namespace.EventName'`

This seems to be an edge case when you've got two distinct events both in the same namespace and happen to be that the first event is a prefix of the second one.

The way events are registered (`Events.DerivedEvent, Shared, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null;
Events.BaseEvent, Shared, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null`), one could argue that it would be possible to valuate the 3rd case by including the comma (`,`) in the filter. Changing this behaviour could be a breaking change for those that rely on the original behaviour though.

Are you blocked by this @jgoyvaerts and workaround is not an option?

**Comment by jgoyvaerts** [comment]
Workaround is implemented at the moment, so not blocked by this. However, we ran into it randomly, and while we realise it's an edge case, it still seems either a flaw by design or a bug. No time pressure in fixing it if it's a bug though!

**Comment by yvesgoeleven** [comment]
There is not much we can do about this, the reason of the prepend check is that if the message is part of a hierarchy (inheritance), then all types will be in the string e.g. `Events.ISentMail, Events.BaseSendMail, Events.ConcreteSendMail`. So an exact match would never match. Therefore we cannot differentiate this from `Events.ISentMailWithDocument`.

**Comment by jgoyvaerts** [comment]
If the decision is to leave it like it is (either due to risking a breaking change or because it's too hard to fix), how about implementing a warning that you have an event which matches another rule? Shouldn't be too hard to detect and it would have saved us (and hopefully others) some precious time in figuring out why our endpoints were spouting "no handler found" exceptions

**Comment by SeanFeldman** [comment]
That's not exactly how it works @jgoyvaerts.
We create for each event subscription/rule and it's a very slow process. To validate that an event could be captured by something else, it would slow things even more and not always capture the problem. Imaging `ISentMail` registered first and then `ISentMailWithDocument` registration coming later. It would not know that a rule for `ISentMail` would satisfy `ISentMailWithDocument` events as well.

Saying that, I have a theory we should probably validate prior to jumping into conclusions or solutions 🙂. I'll raise an exploratary PR to validate it.

**Comment by SeanFeldman** [comment]
Opened a spike PR to demonstrate the issue.

**Comment by SeanFeldman** [comment]
I've looked into this a bit more, and here's an update.

The problem is [polymorphic events](https://docs.particular.net/samples/azure/polymorphic-events-asb), where multiple types require `LIKE` statements with variations.
Another option to solve this issue is to break apart complex SQL filter into multiple Correlation filters.
Given that multiple rules are evaluated using `OR` condition, polymorphic events would still work and "accidental" catch all side effect would be addressed as equation only based filter would take place (`[NServiceBus.EnclosedMessageTypes] EQ 'Events.ISentMail'`).

**Pros**:
- Simpler rules.
- Improved speed (correclation filtering is `O(n)`).
- Increased number of messages (maximum 100K rules per topic vs 2K SQL filters).

**Cons**:
- An additional transport specific header would be required for filtering that is specific to the most derived type rather than the generic `NServiceBus.EnclosedMessageTypes` header.

This is probably going to be a solid candidate for the ASB .NET Standard transport.

**Comment by yvesgoeleven** [comment]
> An additional transport specific header would be required for filtering that is specific to the most derived type rather than the generic NServiceBus.EnclosedMessageTypes header.

I think you need one for every type in the hierarchy, otherwise you can't subscribe to base types or interfaces. Speaking of which, it's possible to send dynamic types generated for an interface and those won't match if there would only be a header for that type.

**Comment by SeanFeldman** [comment]
> I think you need one for every type in the hierarchy

Assuming there's a derived type
```c#
interface ISentMailWithDocument2 : ISentMailWithDocument  {}
```

Subscriber would have to create rules based on what it's interested in:

There would be two rules generated
Rule1: `[NServiceBus.EnclosedMessageTypes] = 'ISentMailWithDocument'`
Rule2: `[NServiceBus.EnclosedMessageTypes] = 'ISentMailWithDocument2'`
Rule3: `[NServiceBus.EnclosedMessageTypes] = 'ISentMailWithDocument1, ISentMailWithDocument2'`

**Comment by yvesgoeleven** [comment]
@SeanFeldman or [NServiceBus.EnclosedMessageTypes] = 'ISentMailWithDocument2, ISentMailWithDocument1' or [NServiceBus.EnclosedMessageTypes] = 'SomeDynamic<T>, ISentMailWithDocument2'...

Order is not determined, neither is implementation type

**Comment by SeanFeldman** [comment]
Where's `SomeDynamic` part is coming from?

**Comment by yvesgoeleven** [comment]
@SeanFeldman this feature: https://docs.particular.net/nservicebus/messaging/messages-as-interfaces. It's allowed to do send<IMyInterface> without there being an implementation for said interface. A type is generated on the fly.

**Comment by SeanFeldman** [comment]
While an underlying implementation of the event interface has some class, that class is **not** listed in the `NServiceBus.EnclosedMessageTypes` NServiceBus header. Only the contracts/specific classes. 

Are you implying that's not the case @yvesgoeleven ?

**Comment by SeanFeldman** [comment]
For the following hierarchy 
```c#
public interface IBaseOrderReceived : IEvent { }

public interface IOrderReceived : IBaseOrderReceived
{
    Guid OrderId { get; set; }
}
```

Generated NServiceBus header will look like the following

```json
"NServiceBus.EnclosedMessageTypes": "IOrderReceived, Shared, Version=1.0.0.0, Culture=neutral, 
PublicKeyToken=null;IBaseOrderReceived, Shared, Version=1.0.0.0, Culture=neutral, 
PublicKeyToken=null"
```

There are two options to subscribe

**Options 1**

`IHandleMessages<IOrderReceived>`

For this option, only messages of `IOrderReceived` type should be subscribed to.

**Option 2**

`IHandleMessages<IBaseOrderReceived>`

For this option, messages of `IOrderReceived` and `IBaseOrderReceived` type should be subscribed to.

**What does it mean for subscribers?**

When subscriber is creating a subscription, it will query to find out all the extenders for an event it subscribes to.

For `IOrderReceived` there are no extenders. Therefore the only rule that should be created is
`Rule1: NServiceBus.EnclosedMessageTypes-ext = IOrderReceived`

For `IBaseOrderReceived` there are no extenders, `IBaseOrderReceived`. Therefore an additional rule should be created resulting in total two rules.
`Rule1: NServiceBus.EnclosedMessageTypes-ext1 = IOrderReceived` // extender
`Rule2: NServiceBus.EnclosedMessageTypes-ext2 = IBaseOrderReceived` // actual event we subscribe to

And that's where it breaks. We can't do it with a single additional header.
But...

We could offer an optimization when an event is not a polymorphic event. I.e. if it's a simple event w/o entenders, we'll create a single rule with Correlation filter. Otherwise, we'll create a single rule with SQL filter.

(unfortunately, it won't address this issue 😊)

**Comment by yvesgoeleven** [comment]
> Are you implying that's not the case

IIRC it used to be different

**Comment by yvesgoeleven** [comment]
> We could offer an optimization when an event is not a polymorphic event. I.e. if it's a simple event w/o entenders, we'll create a single rule with Correlation filter. Otherwise, we'll create a single rule with SQL filter.

sounds good, should get the perf improvements for majority of subscribes then

**Comment by SeanFeldman** [comment]
Solution for greedy problem would be to modify the rules with `[NServiceBus.EnclosedMessageTypes] LIKE '%EventFullName%,'`. The comman will present either there's a single type or there's an inheritance chain.

**Comment by yvesgoeleven** [comment]
@SeanFeldman as discussed in the call, there is the ducktyping case where the `,` is not going to be present

**Comment by SeanFeldman** [comment]
After lengthy internal debates, we could go with the comma <kbd>,</kbd> approach, but it would break the native iterop, which relies on duck-typing.

Looking it pros and cons, a greedy match would result in a logged exception that a handler for an event is not found. This event would not be subscribed to by the endpoint anyways, therefore there's no message loss. With the comma approach, if the native sender would send a FullType only w/o the comma at the end, this would result in lost messages under subscription.

Therefore, closing this issue as `won't fix` and raising a separate issue #717 to simplify subscription rule and improve its performance.
